### PR TITLE
Refactor connection status as an enum

### PR DESCRIPTION
**Pull request by lickel**
We noticed a hole where our socket was getting disconnected and not reconnecting. 
In tracing down code paths, we noticed that the `isConnecting` and `didDisconnect` flags aren't currently thread safe. Further, only the `isConnected` state is exposed to consumers.

This exposes the underlying status as an enum, and guards access behind the same `mutex` that `connected` previously used.

### CODE REVIEW

File: Sources/Starscream/WebSocket.swift [hunk 1]
```diff
@@ -57,6 +57,12 @@ public struct WSError: Error {
     public let code: Int
 }
 
+public enum WebSocketStatus {
```
Comment (fassko) [inline Sources/Starscream/WebSocket.swift]:
```suggestion
public enum WebSocketStatus: String {
```

How about using String instead of default Int?

Comment (lickel) [reply Sources/Starscream/WebSocket.swift]: By default, Swift enums aren’t backed by anything. The default isn’t raw representable at all. 

With that in mind, do you still want this change?

**Comment by lickel** [comment]
@fassko I left a comment above. Mind taking a look?

**Comment by acmacalister** [comment]
The codebase has actually under gone a major refactor and completely changed how we handle the whole isConnected state. Feel free to check it out here #653. Feedback welcomed.

**Comment by daltoniam** [comment]
I'm going to go ahead and close this out as it is related to the old 3.X.X codebase. The connection state should be handled properly in the 4.0.0 codebase. Thanks for the PR!

**Pull request closed without merging**